engine/src/main/java/org/camunda/bpm/engine/impl/persistence/entity/VariableInstanceEntity.java

```diff
@@ -278,26 +278,52 @@ public TypedValue getTypedValue(boolean deserializeValue) {
     return typedValueField.getTypedValue(deserializeValue, isTransient);
   }
 
-  public void setValue(TypedValue value) {
-    // clear value fields
-    clearValueFields();
+  public void setValue(TypedValue newValue) {
+    clearValueFieldsBasedOnType(newValue);
 
-    typedValueField.setValue(value);
+    typedValueField.setValue(newValue);
+  }
+
+  public void clearValueFieldsBasedOnType(TypedValue newValue) {
+    if (variableWillChangeType(newValue)) {
+      clearValueFields(true);
+    } else {
+      clearValueFields(false);
+    }
   }
 
-  public void clearValueFields() {
+  public void clearValueFields(boolean deleteVariable) {
     this.longValue = null;
     this.doubleValue = null;
     this.textValue = null;
     this.textValue2 = null;
     typedValueField.clear();
 
-    if(byteArrayField.getByteArrayId() != null) {
+    if (deleteVariable && byteArrayField.getByteArrayId() != null) {
       deleteByteArrayValue();
       setByteArrayValueId(null);
     }
   }
 
+  public boolean variableWillChangeType(TypedValue newValue) {
+    String currentTypeName = typedValueField.getTypeName();
+    String newTypeName = getTypeName(newValue);
+
+    return !currentTypeName.equals(newTypeName);
+  }
+
+  public String getTypeName(TypedValue value) {
+    if (value.getType() != null) {
+      return value.getType().getName();
+    }
+
+    // Fetches the class name from the value if the type is null
+    return value.getValue()
+            .getClass()
+            .getSimpleName()
+            .toLowerCase();
```

Comment:
❓ In which cases is the type of the variable null?

Reply:
This code is outdated after the latest changes to avoid deserializing the variable during the type fetching.